pageserver/src/tenant/timeline.rs

```diff
@@ -3704,81 +3370,47 @@ impl Timeline {
-        let mut remove_layers = Vec::new();
 
         for l in new_layers {
-            let new_delta_path = l.path();
-
-            let metadata = new_delta_path.metadata().with_context(|| {
-                format!(
-                    "read file metadata for new created layer {}",
-                    new_delta_path.display()
-                )
-            })?;
-
             if let Some(remote_client) = &self.remote_client {
-                remote_client.schedule_layer_file_upload(
-                    &l.filename(),
-                    &LayerFileMetadata::new(metadata.len()),
-                )?;
+                let m = LayerFileMetadata::new(l.layer_desc().file_size);
+                // upload even if duplicated, because we may have changed the contents
+                remote_client.schedule_layer_file_upload(l.clone(), &m)?;
             }
-
-            // update the timeline's physical size
-            self.metrics
-                .resident_physical_size_gauge
-                .add(metadata.len());
-
-            new_layer_paths.insert(new_delta_path, LayerFileMetadata::new(metadata.len()));
-            l.access_stats().record_residence_event(
-                LayerResidenceStatus::Resident,
-                LayerResidenceEventReason::LayerCreate,
-            );
-            let l = l as Arc<dyn PersistentLayer>;
-            if guard.contains(&l) {
+            if guard.contains(l.as_ref()) {
                 duplicated_layers.insert(l.layer_desc().key());
+            } else if LayerMap::is_l0(l.layer_desc()) {
+                return Err(CompactionError::Other(anyhow!("compaction generates a L0 layer file as output, which will cause infinite compaction.")));
             } else {
-                if LayerMap::is_l0(l.layer_desc()) {
-                    return Err(CompactionError::Other(anyhow!("compaction generates a L0 layer file as output, which will cause infinite compaction.")));
-                }
                 insert_layers.push(l);
             }
         }
 
-        // Now that we have reshuffled the data to set of new delta layers, we can
-        // delete the old ones
-        let mut layer_names_to_delete = Vec::with_capacity(deltas_to_compact.len());
-        for ldesc in deltas_to_compact {
-            if duplicated_layers.contains(&ldesc.key()) {
-                // skip duplicated layers, they will not be removed; we have already overwritten them
-                // with new layers in the compaction phase 1.
-                continue;
-            }
-            layer_names_to_delete.push(ldesc.filename());
-            remove_layers.push(guard.get_from_desc(&ldesc));
-        }
+        let remove_layers = {
+            let mut deltas_to_compact = deltas_to_compact;
+            // only remove those inputs which were not outputs
+            deltas_to_compact.retain(|l| !duplicated_layers.contains(&l.layer_desc().key()));
+            deltas_to_compact
+        };
```

Comment:
nit: would have preferred you avoid the conversion to `retain` from this PR; not a huge distraction from the main theme of this PR, but, small distractions add up.

Also, I think the original comments on this paragraph of code were better / more explict than your new one.

Reply:
Which comments were better? How far along upwards are you commenting on this?

Reply:
The ones that this removed.

```
   // Now that we have reshuffled the data to set of new delta layers, we can
        // delete the old ones
        let mut layer_names_to_delete = Vec::with_capacity(deltas_to_compact.len());
        for ldesc in deltas_to_compact {
            if duplicated_layers.contains(&ldesc.key()) {
                // skip duplicated layers, they will not be removed; we have already overwritten them
                // with new layers in the compaction phase 1.
                continue;
            }
            layer_names_to_delete.push(ldesc.filename());
            remove_layers.push(guard.get_from_desc(&ldesc));
        }
```

Reply:
Yes there was so many distractions in the code like `Arc<PersistentLayerDesc>` living only for the duration of the compaction reporting and the other Arc's wildly being cloned in every direction, had a really hard time with this.

We no longer need to collect the ldesc.filename... Okay, perhaps I see what you are suggesting.

I think I need to think about this duplicate layer business more as well. Even though we should have the test case. I would had expected it to start failing if I had done something suprising, but perhaps we don't assert enough or look if the uploads are succeeding with the correct length, or perhaps in the test we always end up creating similar length files.

Reply:
Okay, I think I need to add some machinery for the overwriting support, because now we cannot let the old `Layer` instance control the file at all, otherwise it might get evicted before it's uploaded.

I think the #4696 just left the overlapping layer in layermap, and scheduled upload for the new one, and that was fine, because the layers would had been deleted at the same time by some future gc if this would had ever happened, and we would had just skipped it during eviction (NotFound introduced in some of my PRs quite recently in relation to the same quest).

Luckily we don't have any of these with the GC in logs.

Reply:
We also cannot download such layers, so they need to be made resident before we start l0 => l1 compaction.

I had missed the uploading the new layers and zero test failures happened.

Reply:
Overwriting **layers published to LayerMap** no longer happens with #5172. We might still overwrite files on disk before publishing happens, but that's no issue. In the last commits around https://github.com/neondatabase/neon/pull/4938/commits/54720cf9c586301c769fa992c21c03a1a1ee7a00 I removed error paths near the publishing, so we should not fail there.

Trying to add support for overwritable L1's was too difficult and I decided to implement the crash-consistency via layer map updates instead.